pyvsystems/contract.py

```diff
@@ -1,115 +1,290 @@
+        return DataEntry(struct.unpack(">Q", bytes_object[1:])[0], bytes_object[0:1])
+    elif bytes_object[0:1] == Type.short_bytes:
+        return DataEntry(bytes2str(bytes_object[3:]), bytes_object[0:1])
+
+
+def check_data_type(data, data_type):
+    if data_type == Type.public_key:
+        data_bytes = base58.b58decode(data)
+        return len(data_bytes) == Type.key_length
+    elif data_type == Type.address:
+        data_bytes = base58.b58decode(data)
+        return len(data_bytes) == Type.address_length
+    elif data_type == Type.amount:
+        data_bytes = struct.pack(">Q", data)
+        return len(data_bytes) == Type.amount_length and struct.unpack(">Q", data_bytes)[0] > 0
+    elif data_type == Type.int32:
+        data_bytes = struct.pack(">I", data)
+        return len(data_bytes) == Type.int32_length and struct.unpack(">I", data_bytes)[0] > 0
+    elif data_type == Type.short_text:
+        data_bytes = Deser.serialize_array(str2bytes(data))
+        return struct.unpack(">H", data_bytes[0:2])[0] + 2 == len(data_bytes) and len(
+            data_bytes) <= Type.max_short_text_size + 2
+    elif data_type == Type.short_bytes:
+        data_bytes = Deser.serialize_array(str2bytes(data))
+        return struct.unpack(">H", data_bytes[0:2])[0] + 2 == len(data_bytes) and len(
+            data_bytes) <= Type.max_short_bytes_size + 2
+    else:
+        return True
+
+
+class DataEntry:
+    def __init__(self, data, data_type):
+        if not check_data_type(data, data_type):
+            raise ValueError("Invalid DataEntry data: %s, type: %s" % (str(data), str(data_type)))
+        if data_type == Type.public_key:
+            self.data_bytes = base58.b58decode(data)
+            self.data_type = 'public_key'
+        elif data_type == Type.address:
+            self.data_bytes = base58.b58decode(data)
+            self.data_type = 'address'
+        elif data_type == Type.amount:
+            self.data_bytes = struct.pack(">Q", data)
+            self.data_type = 'amount'
+        elif data_type == Type.int32:
+            self.data_bytes = struct.pack(">I", data)
+            self.data_type = 'int32'
+        elif data_type == Type.short_text:
+            self.data_bytes = Deser.serialize_array(str2bytes(data))
+            self.data_type = 'short_text'
+        elif data_type == Type.contract_account:
+            self.data_bytes = base58.b58decode(data)
+            self.data_type = 'contract_account'
+        elif data_type == Type.token_id:
+            self.data_bytes = base58.b58decode(data)
+            self.data_type = 'token_id'
+        elif data_type == Type.timestamp:
+            self.data_bytes = struct.pack(">Q", data)
+            self.data_type = 'timestamp'
+        elif data_type == Type.short_bytes:
+            self.data_bytes = Deser.serialize_array(str2bytes(data))
+            self.data_type = 'short_bytes'
+        self.data = data
+        self.bytes = data_type + self.data_bytes
+
 
+class Type:
```

Comment:
you may consider using
```python
from typing import NamedTuple
import struct
class TypeStruct(NamedTuple):
  index: bytes
  length: int
  name: str
class Type:
  public_key = TypeStruct(struct.pack(">B", 1), 1, 'public_key')
```
to simplify the code. which may be helpful to the previous coding logic.